src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfJsonPrimitive.cs

```diff
@@ -0,0 +1,406 @@
+    {
+        internal static JsonValue CreatePrimitiveValue(ref Utf8JsonReader reader, JsonNodeOptions options)
+        {
+            switch (reader.TokenType)
+            {
+                case JsonTokenType.False:
+                case JsonTokenType.True:
+                    return new JsonValueOfJsonBool(reader.GetBoolean(), options);
+                case JsonTokenType.String:
+                    byte[] utf8String;
+
+                    if (reader.ValueIsEscaped)
+                    {
+                        ReadOnlySpan<byte> span = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
+                        utf8String = JsonReaderHelper.GetUnescaped(span);
+                    }
+                    else
+                    {
+                        utf8String = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan.ToArray();
+                    }
+
+                    return new JsonValueOfJsonString(utf8String, options);
+                case JsonTokenType.Number:
+                    byte[] numberValue = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan.ToArray();
+                    return new JsonValueOfJsonNumber(new JsonValueOfJsonNumber.JsonNumber(numberValue), options);
+                default:
+                    Debug.Fail("Only primitives allowed.");
+                    ThrowHelper.ThrowJsonException();
+                    return null!; // Unreachable, but required for compilation.
+            }
+        }
+
+        private sealed class JsonValueOfJsonString : JsonValue
+        {
+            private readonly byte[] _value;
+
+            internal JsonValueOfJsonString(byte[] utf8String, JsonNodeOptions? options)
+                : base(options)
+            {
+                _value = utf8String;
+            }
+
+            internal override JsonNode DeepCloneCore() => new JsonValueOfJsonString(_value, Options);
+            private protected override JsonValueKind GetValueKindCore() => JsonValueKind.String;
+
+            public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? options = null)
+            {
+                ArgumentNullException.ThrowIfNull(writer);
+
+                writer.WriteStringValue(_value);
+            }
+
+            public override T GetValue<T>()
+            {
+                if (!TryGetValue(out T? value))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_NodeUnableToConvertElement(JsonValueKind.String, typeof(T));
+                }
+
+                return value;
+            }
+
+            public override bool TryGetValue<T>([NotNullWhen(true)] out T? value)
+                where T : default
+            {
+                if (typeof(T) == typeof(JsonElement))
+                {
+                    int firstByteToEscape = JsonWriterHelper.NeedsEscaping(_value, JavaScriptEncoder.Default);
```

Comment:
Would it be possible to hide this logic and the relevant helper methods behind `JsonWriterHelper` or `JsonDocument`? It seems somewhat out of place in this layer.